src/openassetio-python/tests/bridge/python/test_converter.cpp

```diff
@@ -0,0 +1,268 @@
+    const TraitsDataPtr traitsData = TraitsData::make();
+    PyObject* pyTraitsData = openassetio::python::converter::castToPyObject(traitsData);
+    REQUIRE(pyTraitsData != nullptr);
+
+    WHEN("data is set via the C++ object") {
+      traitsData->addTrait(kTestTraitId);
+
+      THEN("Python object reflects that data set") {
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+
+    WHEN("data is set via the Python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+
+      THEN("C++ object reflects the data set") { CHECK(traitsData->hasTrait(kTestTraitId)); }
+    }
+    Py_DECREF(pyTraitsData);
+  }
+
+  GIVEN("A Python object casted to a C++ object") {
+    const py::object pyClass = py::module_::import("openassetio").attr("TraitsData");
+    py::object pyInstance = pyClass();
+    PyObject* pyTraitsData = pyInstance.release().ptr();
+    const TraitsDataPtr traitsData =
+        openassetio::python::converter::castFromPyObject<TraitsData>(pyTraitsData);
+
+    WHEN("data is set via the C++ object") {
+      traitsData->addTrait(kTestTraitId);
+
+      THEN("Python object reflects that data set") {
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+
+    WHEN("data is set via the Python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+      THEN("C++ object reflects the data set") { CHECK(traitsData->hasTrait(kTestTraitId)); }
+    }
+    Py_DECREF(pyTraitsData);
+  }
+}
+
+SCENARIO("Casting to PyObject extends object lifetime") {
+  GIVEN("A Python object casted from a C++ object") {
+    TraitsDataPtr traitsData = TraitsData::make();
+    traitsData->addTrait(kTestTraitId);
+    PyObject* pyTraitsData = openassetio::python::converter::castToPyObject(traitsData);
+
+    WHEN("C++ reference is destroyed") {
+      CHECK(Py_REFCNT(pyTraitsData) == 1);  // Initial condition.
+      traitsData.reset();
+
+      THEN("object remains alive and can be operated on via the Python interpreter") {
+        CHECK(Py_REFCNT(pyTraitsData) == 1);
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+      Py_DECREF(pyTraitsData);
```

Comment:
[question] Should this be up one level in the scope, as the incref happens at 99?

Reply:
FWIW, is it worth just using the pybind object to save this kind of thing even coming up?

Reply:
Yeah, although not technically a different call-order here because there's only one case, this'll break if someone adds another leaf without thinking.

> is it worth just using the pybind object to save this kind of thing even coming up?
I thought, and still think, that it's worth dogfooding the way people will need to interact with our methods when we write our tests, otherwise we don't have any context about the burdens of using these interfaces as we provide them. Since the whole point is to not have pybind, I think it's good that we put ourselves through it.

On the other hand, one could argue test stability is more important than my abstract gesture at building design-sense...

Reply:
> Since the whole point is to not have pybind,

Don't know if its relevant, but it's perhaps more to hide _our_ use of pybind. I'd imagine most calling contexts will use some management layer (pybind or other), so in a way having the tests use pybind and having to interact with our raw CPython object is actually more representative. I think most people avoid raw CPython where they can due to 🦶 🔫 🔫 🔫

Reply:
3e2f4b05e9126bbb58a95b61593fdd2a7fa0412b

As per our discussion, simply fixed the existing formulation via moving the DECREF up a scope.